atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresDdlTable.java

```diff
@@ -118,10 +124,56 @@ public void checkDatabaseVersion() {
 
     @Override
     public void compactInternally() {
+        if (config.compactIntervalMillis() > 0 && checkIfTableHasNotBeenCompactedForCompactIntervalMillis()) {
+            runCompactOnTable();
+        }
+    }
+
+    @VisibleForTesting
+    boolean checkIfTableHasNotBeenCompactedForCompactIntervalMillis() {
+        return getCurrentDatabaseTimestamp().getTime() - getLastVacuumTimestamp().getTime()
+                > config.compactIntervalMillis();
+    }
+
+    private Timestamp getLastVacuumTimestamp() {
+        AgnosticResultSet vaccumTimestamps = conns.get().selectResultSetUnregisteredQuery(
+                "SELECT relname, "
+                        + "last_vacuum, "
+                        + "last_autovacuum, "
+                        + "last_analyze, "
+                        + "last_autoanalyze "
+                        + "from pg_stat_user_tables where relname = ?",
+                prefixedTableName());
+
+        AgnosticResultRow vacuumTimestampsForTable = Iterables.getOnlyElement(vaccumTimestamps.rows());
+
+        return ImmutableList.of("last_vacuum", "last_autovacuum", "last_analyze", "last_autoanalyze")
+                .stream()
+                .map(str -> getTimestamp(vacuumTimestampsForTable.getObject(str)))
+                .max(Timestamp::compareTo)
+                .orElse(new Timestamp(0));
+    }
+
+    private Timestamp getCurrentDatabaseTimestamp() {
+        AgnosticResultSet currentTimestamp = conns.get().selectResultSetUnregisteredQuery("SELECT CURRENT_TIMESTAMP");
+        return getTimestamp(Iterables.getOnlyElement(currentTimestamp.rows()).getObject("now"));
+    }
+
+    private void runCompactOnTable() {
         // VACUUM FULL is /really/ what we want here, but it takes out a table lock
         conns.get().executeUnregisteredQuery("VACUUM ANALYZE " + prefixedTableName());
     }
 
+    private Timestamp getTimestamp(Object timestampFromResultSet) {
+        if (timestampFromResultSet == null) {
+            return new Timestamp(0);
+        } else {
+            Preconditions.checkState(timestampFromResultSet instanceof Timestamp,
```

Comment:
Should we throw here or just log and return Timestamp(0)? If somehow this actually fails, we may end up in a state where we will never vacuum.

Reply:
I think this should be a guarantee of the resultset, we have an ETE test that covers this.